### DISCUSSION TITLE
Populate item form fields with data received from external service

### DISCUSSION THREAD

**Discussion by tmdoit** [directus/directus#15517] [Archive]
Hi! 
I'm trying to choose between Budibase and Directus, one functionality which I need is the possibility to populate item form with data received from  the external service, I know that such option is not provided from the box, so I'm trying to figure out how hard it will be to implement this. 
![Zrzut ekranu 2022-09-11 o 09 34 51](https://user-images.githubusercontent.com/85706044/189519708-ad39b744-f608-4510-9d23-77af82f2518b.png)

I'm confused with Directus extending possibility, can you please describe in short the process of implementing discussed functionality?
1. Where/How to add button
2. Where/How to make a request
3. How to populate item form fields with the response received from the external service

Thanks in advance! :)

**Comment by azrikahar** [comment]
This use case should be possible with [Flows](https://docs.directus.io/configuration/flows.html)! Here's a quick example:

1. Create a data model partially based on https://jsonplaceholder.typicode.com/users/1 (only some of the fields for demo purpose)
2. Create a Flow with Manual trigger, and setting it `Item Page Only` so it will only appear on the sidebar for a particular item 
    - this should address your first question
3. In the same flow, add a Webhook / Request URL operation and fetch from `https://jsonplaceholder.typicode.com/users/{{$trigger.body.keys[0]}}` (notice that we use `keys[0]`, because by default it is sent as an array for use cases such as multiple items. But since we're only triggering this for a specific item, we can use `[0]` to grab the first primary key which is the current item's ID. Also make sure to set the operation Key into something recognizable such as `external_req` so that we can refer to it's data in other operations.
    - this should address your second question
4. In the same flow after the request URL operation, add an Update Data operation and target the same collection, full permissions (this is just in case the current user does _not_ have update permission, depending on your workflow), IDs set as the same array from `$trigger.body.keys`, payload mirrors the data model we created in step 1. However, here we will "fill" in the data with the API result in the `external_req` operation!
    - this should address your third question
5. Go to the item (I've created it prior with empty data so that we can have the ID `1`) sidebar, and run the flow. We should see the form being updated with information from https://jsonplaceholder.typicode.com/users/1.

## Quick video for above example

https://user-images.githubusercontent.com/42867097/189567391-2f066477-6459-4d59-adb2-47b9e53d9457.mp4

**Reply by tmdoit** [reply]
It's a workaround because of bad UX, the information received from the external service is needed to create a record, not to update it. Second thing is that if you modify ie. "name" field and after you will run the flow, field value will not be updated/replaced with the value returned by the flow.
Thank you for your response, I wasn't aware of such possibility. 

Can we please continue with my questions from previous message regarding the implementation such functionality by extending Directus?

**Reply by azrikahar** [reply]
> It's a workaround because of bad UX, the information received from the external service is needed to create a record, not to update it.

Understandably it's not ideal, but my initial impression was that it's used to populate the current form 🙂

> Second thing is that if you modify ie. "name" field and after you will run the flow, field value will not be updated/replaced with the value returned by the flow.

This is just based on the example above. If you use the Run Script operation for example, you would be able to have finer control on what values to update/replace or even use lodash's `merge()` to merge the values. 

> Can we please continue with my questions from previous message regarding the implementation such functionality by extending Directus?

There isn't exactly a direct way to add a button beyond the button links, but you will need to create your own [interface](https://docs.directus.io/extensions/interfaces.html) or [display](https://docs.directus.io/extensions/displays.html) depending on your use case and add your custom logic in your component. You can then use the `useApi()` composable for example to make requests: https://docs.directus.io/extensions/displays.html#accessing-internal-systems. You can also update the value of other fields by emitting the `setFieldValue` event as described here: https://github.com/directus/directus/pull/11620#issuecomment-1141053792

**Reply by tmdoit** [reply]
> This is just based on the example above. If you use the Run Script operation for example, you would be able to have finer control on what values to update/replace or even use lodash's merge() to merge the values.

Sorry, I meant when user will change value in input field "name" via item form, form input field "name" will be not updated by the flow - it's expected behavior or a bug?

**Reply by azrikahar** [reply]
Oh I believe you mean what was discussed over here https://github.com/directus/directus/pull/14700#discussion_r931247926 right? In that case, it is indeed the expected behavior for now since it's technically still a staged but unsaved value. It might be possible to be improved on once there's websockets support.

**Reply by tmdoit** [reply]
Exactly, it's clear to me now.
Many thanks for your help! 🍻

**Reply by mikgross** [reply]
@azrikahar thanks for the nice demo, this looks nice.
I want to do something different and I can't figure how to.

I would like to have a dropdown field with external data from an API, without having to sync this data in my DB => query it every time I enter a form.

Then my user should be able to select the data they need and I to save a partial copy of it on that particular field.

I am suspecting I can't do this without saving the data on the DB. Is that correct?

**Comment by ivanakruljac** [comment]
I want to trigger the API with filed value.
For example, I have payload filed with name **external_id**.
Trigger should be field value, not primary key value.
So, I want to use it like this https://www.myapi.com/profiles/{{$trigger.body.external_id}}
I need **external_id** value to be entered manually.
